lib/mariaex/protocol.ex

```diff
@@ -598,25 +598,51 @@ defmodule Mariaex.Protocol do
     sock_mod.send(sock, data)
   end
 
-  defp msg_recv(%{sock: sock, state: state, timeout: timeout}) do
-    msg_recv(sock, state, timeout)
+  defp msg_recv(%__MODULE__{sock: sock_info, buffer: buffer}=state) do
+    msg_recv(sock_info, state, buffer)
   end
 
-  defp msg_recv({sock_mod, sock}, decode_state, timeout) do
-    case sock_mod.recv(sock, 4, timeout) do
-      {:ok, << len :: size(24)-little-integer, _seqnum :: size(8)-integer >> = header} ->
-        case sock_mod.recv(sock, len, timeout) do
-          {:ok, packet_body} ->
-            {packet, ""} = decode(header <> packet_body, decode_state)
-            {:ok, packet}
-          {:error, _} = error ->
-            error
-        end
-      {:error, _} = error ->
-        error
+  defp msg_recv(sock, state, buffer) do
+    case msg_decode(buffer, state) do
+      {:ok, _packet, _new_state}=success ->
+        success
+
+      {:more, more} ->
+        msg_recv(sock, state, buffer, more)
+
+      {:error, _}=err ->
+        err
     end
   end
 
+  defp msg_recv({sock_mod, sock}=s, state, buffer, more) do
+
+    case sock_mod.recv(sock, more, state.timeout) do
+      {:ok, data} when byte_size(data) < more ->
+        msg_recv(s, state, [buffer | data], more - byte_size(data))
+
+      {:ok, data} when is_binary(buffer) ->
+        msg_recv(s, state, buffer <> data)
+
+      {:ok, data} when is_list(buffer) ->
+        msg_recv(s, state, IO.iodata_to_binary([buffer | data]))
+
+      {:error, _} = err ->
+          err
+    end
+  end
+
+
+  def msg_decode(<< len :: size(24)-little-integer, _seqnum :: size(8)-integer, message :: binary>>=header, state) when byte_size(message) >= len do
+
+    {packet, rest} = decode(header, state.state)
```

Comment:
We can call `decode/2` directly from `msg_recv/3` to prevent the double decode, this function is doing exactly the same thing isn't it?